File: python/nexo_plugin_sdk/adapter.py
"""Child-side dispatch loop.

Mirrors the Rust counterpart in
``crates/microapp-sdk/src/plugin.rs::PluginAdapter``. Reads JSON-RPC
2.0 newline-delimited frames from stdin and dispatches:

  - ``method == "initialize"`` (request) → reply with manifest +
    server_version.
  - ``method == "broker.event"`` (notification) → spawn a detached
    task running ``on_event`` so the reader keeps polling stdin while
    the handler awaits its own broker / host interactions.
  - ``method == "shutdown"`` (request) → abandon in-flight host calls,
    drain in-flight handler tasks, reply ``{"ok": true}``, exit.
  - ``method == "llm.complete.delta"`` (notification) → route the chunk
    to the awaiting :class:`~nexo_plugin_sdk.host.LlmStream`.
  - a frame with an ``id`` and no ``method`` → a *response* to a
    child→host request we issued (``memory.recall`` / ``llm.complete``)
    → resolve the awaiting future; unknown id → dropped with a warn.
  - anything else with an ``id`` → reply ``-32601 method not found``.
  - anything else without an ``id`` → silently ignored (JSON-RPC §4.1).

Wire format: ``nexo-plugin-contract.md``.
"""

import asyncio
import itertools
import json
import signal
import sys
from typing import Any, Awaitable, Callable, Iterator, Union

from . import stdout_guard, wire
from .broker import BrokerSender
from .errors import PluginError, RpcServerError, RpcTransportError, WireError
from .events import Event
from .host import LlmCompleteResult, _STREAM_END, _StreamPending
from .manifest import read_manifest

EventHandler = Callable[[str, Event, BrokerSender], Awaitable[None]]
ShutdownHandler = Callable[[], Awaitable[None]]

_Pending = Union["asyncio.Future[Any]", _StreamPending]

# Signals that trigger a graceful shutdown (drain in-flight handlers,
# then exit 0). Mirrors the TS SDK's SIGTERM/SIGINT handling.
_SHUTDOWN_SIGNALS = (signal.SIGTERM, signal.SIGINT)


def _safe_remove_asyncio_signal(loop: asyncio.AbstractEventLoop, sig: int) -> None:
    try:
        loop.remove_signal_handler(sig)
    except (ValueError, RuntimeError):
        pass


def _safe_restore_signal(sig: int, previous: Any) -> None:
    # signal.signal returns SIG_DFL/SIG_IGN/a callable, or None when a
    # non-Python handler was installed (which we cannot restore).
    if previous is None:
        return
    try:
        signal.signal(sig, previous)
    except (ValueError, OSError, TypeError):
        pass


class PluginAdapter:
    """Wraps the JSON-RPC dispatch loop. Construct once with the
    manifest TOML body, then ``await adapter.run()`` from your async
    entrypoint.
    """

    def __init__(
        self,
        *,
        manifest_toml: str,
        server_version: str = "0.1.0",
        on_event: EventHandler | None = None,
        on_shutdown: ShutdownHandler | None = None,
        enable_stdout_guard: bool = True,
        max_frame_bytes: int = wire.MAX_FRAME_BYTES,
        handle_process_signals: bool = True,
    ) -> None:
        # Parse + validate the manifest first — a failed construction
        # must not leave a dangling stdout guard installed.
        self._manifest = read_manifest(manifest_toml)
        self._server_version = server_version
        self._on_event = on_event
        self._on_shutdown = on_shutdown
        self._enable_stdout_guard = enable_stdout_guard
        self._max_frame_bytes = max_frame_bytes
        self._handle_process_signals = handle_process_signals
        # Single-shot guard — run() may be called at most once.
        self._started = False

        if enable_stdout_guard:
            stdout_guard.install_stdout_guard()
        # Blessed frames (initialize/shutdown replies, broker.publish,
        # child→host requests) write through the captured *original*
        # stdout so they bypass the guard.
        self._stdout = stdout_guard.original_stdout() or sys.stdout

        self._write_lock = asyncio.Lock()
        # Pending child→host requests, keyed by SDK-assigned request id;
        # shared with the broker handle so the dispatch loop can route
        # responses + llm.complete.delta chunks back to awaiting callers.
        self._pending: dict[int, _Pending] = {}
        self._id_alloc: Iterator[int] = itertools.count(1)
        self._broker = BrokerSender(self._write_lock, self._stdout, self._pending, self._id_alloc)
        # In-flight handler tasks — shutdown awaits them before exiting.
        self._inflight: set[asyncio.Task[None]] = set()

    @property
    def manifest(self) -> dict[str, Any]:
        return self._manifest

    def _install_signal_handlers(
        self, loop: asyncio.AbstractEventLoop, stop_event: asyncio.Event
    ) -> list[Callable[[], None]]:
        cleanups: list[Callable[[], None]] = []
        if not self._handle_process_signals:
            return cleanups
        for sig in _SHUTDOWN_SIGNALS:
            try:
                loop.add_signal_handler(sig, stop_event.set)
                cleanups.append((lambda s: lambda: _safe_remove_asyncio_signal(loop, s))(sig))
            except (NotImplementedError, RuntimeError, ValueError):
                try:
                    previous = signal.signal(
                        sig, lambda *_a: loop.call_soon_threadsafe(stop_event.set)
                    )
                except (ValueError, OSError, RuntimeError):
                    sys.stderr.write(
                        f"plugin: signal {sig} not installable on this platform; "
                        "graceful shutdown on signal disabled\n"
                    )
                    sys.stderr.flush()
                    continue
                cleanups.append((lambda s, p: lambda: _safe_restore_signal(s, p))(sig, previous))
        return cleanups

    async def run(self) -> None:
        if self._started:
            raise PluginError("PluginAdapter.run() already invoked")
        self._started = True

        loop = asyncio.get_running_loop()
        reader = asyncio.StreamReader(limit=self._max_frame_bytes + 64 * 1024)
        protocol = asyncio.StreamReaderProtocol(reader)
        transport, _ = await loop.connect_read_pipe(lambda: protocol, sys.stdin)

        stop_event = asyncio.Event()
        signal_cleanups = self._install_signal_handlers(loop, stop_event)
        try:
            while not stop_event.is_set():
                read_task = asyncio.ensure_future(reader.readline())
                stop_task = asyncio.ensure_future(stop_event.wait())
                try:
                    await asyncio.wait({read_task, stop_task}, return_when=asyncio.FIRST_COMPLETED)
                finally:
                    if not read_task.done():
                        read_task.cancel()
                    if not stop_task.done():
                        stop_task.cancel()
                if stop_event.is_set():
                    break
                try:
                    line_bytes = read_task.result()
                except ValueError as e:
                    err = WireError(f"oversized inbound frame dropped: {e}")
                    sys.stderr.write(f"plugin: {err}\n")
                    sys.stderr.flush()
                    continue
                except asyncio.CancelledError:
                    break
                if not line_bytes:
                    break  # EOF — host closed stdin.
                if len(line_bytes) > self._max_frame_bytes:
                    err = WireError(
                        f"inbound frame {len(line_bytes)} bytes exceeds "
                        f"max_frame_bytes {self._max_frame_bytes}"
                    )
                    sys.stderr.write(f"plugin: {err}\n")
                    sys.stderr.flush()
                    continue
                line = line_bytes.decode("utf-8", "replace").strip()
                if not line:
                    continue
                stop = await self._handle_line(line)
                if stop:
                    break
        finally:
            for undo in signal_cleanups:
                undo()
            transport.close()
            self._abandon_pending()
            await self._drain_inflight()
            if self._enable_stdout_guard:
                stdout_guard.uninstall_stdout_guard()

    async def _handle_line(self, line: str) -> bool:
        """Process one JSON-RPC frame. Returns True if the loop should
        stop (a ``shutdown`` request was handled)."""
        try:
            msg = json.loads(line)
        except json.JSONDecodeError as e:
            sys.stderr.write(f"plugin: malformed jsonrpc line: {e}\n")
            sys.stderr.flush()
            return False
        if not isinstance(msg, dict):
            sys.stderr.write("plugin: jsonrpc frame must be an object\n")
            sys.stderr.flush()
            return False
        method = msg.get("method")
        req_id = msg.get("id")

        if method is None:
            # A response (id + result/error) to a request we issued, or
            # garbage. Never reply to it.
            if req_id is not None and req_id in self._pending:
                self._route_response(req_id, msg)
            elif req_id is not None:
                sys.stderr.write(
                    f"plugin: response for unknown/expired request id {req_id!r}, dropped\n"
                )
                sys.stderr.flush()
            return False

        if method == "initialize":
            await self._reply_initialize(req_id)
        elif method == "broker.event":
            params = msg.get("params") or {}
            task = asyncio.create_task(self._dispatch_event(params))
            self._inflight.add(task)
            task.add_done_callback(self._inflight.discard)
        elif method == "llm.complete.delta":
            self._route_delta(msg.get("params") or {})
        elif method == "shutdown":
            self._abandon_pending()
            await self._drain_inflight()
            await self._reply_shutdown(req_id)
            return True
        elif req_id is not None:
            await self._send_error(req_id, -32601, "method not found")
        # Unknown notification (no id) — silently ignore (JSON-RPC §4.1).
        return False

    # ── child→host response routing ───────────────────────────────

    def _route_response(self, req_id: Any, msg: dict[str, Any]) -> None:
        pending = self._pending.pop(req_id, None)
        if pending is None:
            return
        if isinstance(pending, _StreamPending):
            pending.chunks.put_nowait(_STREAM_END)
            if pending.final.done():
                return
            if "error" in msg:
                err = msg["error"] or {}
                pending.final.set_exception(
                    RpcServerError(int(err.get("code", -32603)), str(err.get("message", "")))
                )
            else:
                try:
                    pending.final.set_result(LlmCompleteResult.from_json(msg.get("result") or {}))
                except Exception as e:  # noqa: BLE001
                    pending.final.set_exception(RpcTransportError(f"bad stream final reply: {e}"))
            return
        # Single-shot future.
        if pending.done():
            return
        if "error" in msg:
            err = msg["error"] or {}
            pending.set_exception(
                RpcServerError(int(err.get("code", -32603)), str(err.get("message", "")))
            )
        else:
            pending.set_result(msg.get("result") or {})

    def _route_delta(self, params: dict[str, Any]) -> None:
        rid = params.get("request_id")
        chunk = params.get("chunk")
        pending = self._pending.get(rid)
        if isinstance(pending, _StreamPending) and isinstance(chunk, str):
            pending.chunks.put_nowait(chunk)
        else:
            sys.stderr.write(
                f"plugin: llm.complete.delta for unknown/non-stream request {rid!r}, dropped\n"
            )
            sys.stderr.flush()

    def _abandon_pending(self) -> None:
        """Reject every outstanding child→host request — the host is
        gone (shutdown / EOF), it will not answer them."""
        for req_id in list(self._pending.keys()):
            pending = self._pending.pop(req_id, None)
            if pending is None:
                continue
            if isinstance(pending, _StreamPending):
                pending.chunks.put_nowait(_STREAM_END)
                if not pending.final.done():
                    pending.final.set_exception(RpcTransportError("adapter shutting down"))
            elif not pending.done():
                pending.set_exception(RpcTransportError("adapter shutting down"))

    # ── lifecycle replies ─────────────────────────────────────────

    async def _drain_inflight(self) -> None:
        if not self._inflight:
            return
        await asyncio.gather(*list(self._inflight), return_exceptions=True)

    async def _reply_initialize(self, req_id: Any) -> None:
        await self._send_response(
            req_id, {"manifest": self._manifest, "server_version": self._server_version}
        )

    async def _reply_shutdown(self, req_id: Any) -> None:
        if self._on_shutdown is not None:
            try:
                await self._on_shutdown()
            except Exception as e:  # noqa: BLE001
                sys.stderr.write(f"plugin: on_shutdown raised: {e}\n")
                sys.stderr.flush()
        await self._send_response(req_id, {"ok": True})

    async def _dispatch_event(self, params: dict[str, Any]) -> None:
        if self._on_event is None:
            return
        try:
            topic = params.get("topic")
            raw_event = params.get("event") or {}
            if not isinstance(topic, str):
                raise WireError("broker.event params missing string `topic`")
            if not isinstance(raw_event, dict):
                raise WireError("broker.event params missing dict `event`")
            event = Event.from_json(raw_event)
            await self._on_event(topic, event, self._broker)
        except Exception as e:  # noqa: BLE001
            sys.stderr.write(f"plugin: on_event raised: {e}\n")
            sys.stderr.flush()

    async def _send_response(self, req_id: Any, result: dict[str, Any]) -> None:
        line = wire.serialize_frame(wire.build_response(req_id, result))
        async with self._write_lock:
            self._stdout.write(line)
            self._stdout.flush()

    async def _send_error(self, req_id: Any, code: int, message: str) -> None:
        line = wire.serialize_frame(wire.build_error_response(req_id, code, message))
        async with self._write_lock:
            self._stdout.write(line)
            self._stdout.flush()

File: typescript/src/adapter.ts
/**
 * Child-side dispatch loop.
 *
 * Mirrors the Rust counterpart in
 * `crates/microapp-sdk/src/plugin.rs::PluginAdapter`. Reads JSON-RPC
 * 2.0 newline-delimited frames from stdin, dispatches:
 *
 * - `method == "initialize"` (request) → reply with manifest +
 *   server_version.
 * - `method == "broker.event"` (notification) → spawn a detached task
 *   running `onEvent` so the reader keeps polling stdin while the
 *   handler awaits its own broker / host round-trips.
 * - `method == "shutdown"` (request) → abandon in-flight host calls,
 *   drain in-flight handler tasks, reply `{ok: true}`, exit.
 * - `method == "llm.complete.delta"` (notification) → route the chunk
 *   to the awaiting `LlmStream`.
 * - a frame with an `id` and no `method` → a *response* to a child→host
 *   request we issued (`memory.recall` / `llm.complete`) → resolve the
 *   awaiting promise; unknown id → dropped with a warn.
 * - anything else with an `id` → reply `-32601 method not found`.
 * - anything else without an `id` → silently ignored (JSON-RPC §4.1).
 */

import { Buffer } from "node:buffer";
import * as readline from "node:readline";

import { BrokerSender } from "./broker.js";
import {
  ManifestError,
  PluginError,
  RpcServerError,
  RpcTransportError,
  WireError,
} from "./errors.js";
import { Event } from "./events.js";
import { parseLlmCompleteResult, type Pending } from "./host.js";
import { parseManifest, type ParsedManifest } from "./manifest.js";
import { installStdoutGuard, isStdoutGuardInstalled } from "./stdout-guard.js";
import {
  ToolError,
  ToolExecutionFailedError,
  toolDefToJson,
  type ToolContext,
  type ToolDef,
  type ToolHandler,
  type ToolHandlerWithContext,
  type ToolInvocation,
} from "./tools.js";
import {
  buildErrorResponse,
  buildResponse,
  JSONRPC_VERSION,
  MAX_FRAME_BYTES,
  serializeFrame,
  type JsonRpcErrorResponse,
  type JsonRpcId,
} from "./wire.js";

export type EventHandler = (
  topic: string,
  event: Event,
  broker: BrokerSender,
) => Promise<void>;

export type ShutdownHandler = () => Promise<void>;

/** Return the manifest's `[plugin.extends].tools` list, or `undefined`
 * when the section / field is absent. Throws `ManifestError` when
 * present but not a list of strings. */
function manifestExtendsTools(raw: Record<string, unknown>): string[] | undefined {
  const plugin = raw["plugin"];
  if (typeof plugin !== "object" || plugin === null) return undefined;
  const extends_ = (plugin as Record<string, unknown>)["extends"];
  if (typeof extends_ !== "object" || extends_ === null) return undefined;
  if (!("tools" in extends_)) return undefined;
  const tools = (extends_ as Record<string, unknown>)["tools"];
  if (!Array.isArray(tools) || !tools.every((t) => typeof t === "string")) {
    throw new ManifestError("[plugin.extends].tools must be a list of strings", "plugin.extends.tools");
  }
  return tools as string[];
}

export interface PluginAdapterOptions {
  /** Body of nexo-plugin.toml. Parsed once at construction. */
  manifestToml: string;
  /** Returned in the initialize reply. Default `"0.1.0"`. */
  serverVersion?: string;
  /** Invoked for every broker.event notification. Detached task — the
   * reader does not block while the handler awaits broker / host
   * round-trips. */
  onEvent?: EventHandler;
  /** Awaited before `{ok: true}` reply to shutdown. */
  onShutdown?: ShutdownHandler;
  /** Tool catalog advertised in the `initialize` reply (contract
   * §4.1.1). Every `name` must appear in the manifest's
   * `[plugin.extends].tools` — a name that doesn't is a hard error at
   * `run()` (mirrors the host's drift check). */
  tools?: ToolDef[];
  /** `tool.invoke` dispatch handler — `fn(inv)`, sync or async.
   * Mutually exclusive with `onToolWithContext`; if both are set, the
   * with-context one wins. */
  onTool?: ToolHandler;
  /** `tool.invoke` dispatch handler — `fn(inv, ctx)`, sync or async,
   * where `ctx.broker` lets the tool body call `memoryRecall` /
   * `llmComplete` mid-invocation. Wins over `onTool` when both set. */
  onToolWithContext?: ToolHandlerWithContext;
  /** Default true — patches `process.stdout.write` to divert non-JSON
   * lines to stderr. Critical for plugin authors who accidentally
   * `console.log`. Set false only if you have another guard layer. */
  enableStdoutGuard?: boolean;
  /** Default `MAX_FRAME_BYTES` (1 MiB). Reject inbound frames larger
   * than this with a WireError; dispatch continues. */
  maxFrameBytes?: number;
  /** Default true — listen for SIGTERM + SIGINT and trigger graceful
   * shutdown (drain in-flight, exit 0). */
  handleProcessSignals?: boolean;
}

interface JsonRpcFrameLike {
  jsonrpc?: unknown;
  id?: unknown;
  method?: unknown;
  params?: unknown;
  result?: unknown;
  error?: { code?: unknown; message?: unknown } | undefined;
}

export class PluginAdapter {
  private readonly parsed: ParsedManifest;
  private readonly serverVersion: string;
  private readonly onEvent?: EventHandler;
  private readonly onShutdown?: ShutdownHandler;
  private readonly maxFrameBytes: number;
  private readonly handleProcessSignals: boolean;
  private readonly inflight = new Set<Promise<void>>();
  private readonly broker: BrokerSender;
  private readonly declaredTools: ToolDef[];
  private readonly onTool?: ToolHandler;
  private readonly onToolWithContext?: ToolHandlerWithContext;
  /** Outstanding child→host requests, keyed by SDK-assigned id; shared
   * with the broker handle so the dispatch loop can route responses +
   * `llm.complete.delta` chunks back to awaiting callers. */
  private readonly pending = new Map<number, Pending>();

  private started = false;
  private stopped = false;
  private nextId = 0;
  private rl: readline.Interface | null = null;
  private signalCleanup: (() => void) | null = null;

  constructor(opts: PluginAdapterOptions) {
    this.parsed = parseManifest(opts.manifestToml);
    this.serverVersion = opts.serverVersion ?? "0.1.0";
    this.onEvent = opts.onEvent;
    this.onShutdown = opts.onShutdown;
    this.maxFrameBytes = opts.maxFrameBytes ?? MAX_FRAME_BYTES;
    this.handleProcessSignals = opts.handleProcessSignals ?? true;

    // ── tool dispatch (contract §4.1.1 + §5.t) ──────────────────────
    this.declaredTools = [...(opts.tools ?? [])];
    // with-context wins when both are set — mirrors the Rust SDK.
    this.onToolWithContext = opts.onToolWithContext;
    this.onTool = opts.onToolWithContext !== undefined ? undefined : opts.onTool;
    if (this.declaredTools.length > 0) {
      const allowed = manifestExtendsTools(this.parsed.raw);
      if (allowed === undefined) {
        throw new ManifestError(
          "tools= used but the manifest has no [plugin.extends].tools list; declared: " +
            this.declaredTools.map((t) => t.name).join(", "),
          "plugin.extends.tools",
        );
      }
      const offenders = this.declaredTools.filter((t) => !allowed.includes(t.name)).map((t) => t.name);
      if (offenders.length > 0) {
        throw new ManifestError(
          `declared tool(s) not in [plugin.extends].tools [${allowed.join(", ")}]: ${offenders.join(", ")}`,
          "plugin.extends.tools",
        );
      }
    }

    if (opts.enableStdoutGuard !== false) {
      installStdoutGuard();
    }

    this.broker = new BrokerSender(
      (line) => {
        // Direct stdout write through the original handle — the guard
        // would no-op on JSON lines anyway, but skipping it saves one
        // parse round-trip per frame. The guard remains installed for
        // everything OTHER than the SDK's blessed path.
        process.stdout.write(line);
      },
      this.pending,
      () => ++this.nextId,
    );
  }

  get manifest(): Readonly<Record<string, unknown>> {
    return this.parsed.raw;
  }

  /** Single-shot. Throws PluginError if called twice. */
  async run(): Promise<void> {
    if (this.started) {
      throw new PluginError("PluginAdapter.run() already invoked");
    }
    this.started = true;

    if (this.handleProcessSignals) {
      const onSig = (): void => {
        // Closing the readline interface breaks the for-await loop
        // below; the surrounding code handles drain + exit.
        this.rl?.close();
      };
      process.on("SIGTERM", onSig);
      process.on("SIGINT", onSig);
      this.signalCleanup = (): void => {
        process.removeListener("SIGTERM", onSig);
        process.removeListener("SIGINT", onSig);
      };
    }

    this.rl = readline.createInterface({
      input: process.stdin,
      terminal: false,
      crlfDelay: Infinity,
    });

    try {
      for await (const rawLine of this.rl) {
        if (this.stopped) {
          break;
        }
        await this.handleLine(rawLine);
      }
    } finally {
      this.rl?.close();
      this.rl = null;
      this.signalCleanup?.();
      this.signalCleanup = null;
      // Abandon any outstanding host calls (host is gone), then drain
      // handler tasks — most relevant for SIGTERM-initiated exits.
      this.abandonPending();
      await this.drainInflight();
    }
  }

  private async handleLine(line: string): Promise<void> {
    if (line.length === 0) {
      return;
    }
    const byteLen = Buffer.byteLength(line, "utf-8");
    if (byteLen > this.maxFrameBytes) {
      process.stderr.write(
        `plugin: inbound frame ${byteLen} bytes exceeds maxFrameBytes ${this.maxFrameBytes}\n`,
      );
      return;
    }

    let msg: JsonRpcFrameLike;
    try {
      msg = JSON.parse(line);
    } catch (e) {
      process.stderr.write(`plugin: malformed jsonrpc line: ${e instanceof Error ? e.message : String(e)}\n`);
      return;
    }
    if (typeof msg !== "object" || msg === null) {
      process.stderr.write(`plugin: jsonrpc frame must be an object\n`);
      return;
    }

    const method = msg.method;
    const id = msg.id;

    if (typeof method !== "string") {
      // A response (id + result/error) to a request we issued, or
      // garbage. Never reply to it.
      if (typeof id === "number" && this.pending.has(id)) {
        this.routeResponse(id, msg);
      } else if (id !== undefined && id !== null) {
        process.stderr.write(`plugin: response for unknown/expired request id ${JSON.stringify(id)}, dropped\n`);
      }
      return;
    }

    if (method === "initialize") {
      this.replyInitialize(id);
    } else if (method === "broker.event") {
      this.dispatchEvent(msg.params);
    } else if (method === "tool.invoke") {
      this.handleToolInvoke(id, msg.params);
    } else if (method === "llm.complete.delta") {
      this.routeDelta(msg.params);
    } else if (method === "shutdown") {
      await this.replyShutdown(id);
      this.stopped = true;
      this.rl?.close();
    } else if (id !== undefined) {
      // Unknown request — JSON-RPC requires a reply.
      this.writeFrame(buildErrorResponse(id as never, -32601, "method not found"));
    }
    // Unknown notification (no id) — silently ignore per JSON-RPC §4.1.
  }

  // ── child→host response routing ─────────────────────────────────

  private routeResponse(id: number, msg: JsonRpcFrameLike): void {
    const p = this.pending.get(id);
    if (!p) return;
    this.pending.delete(id);
    const err = msg.error;
    const code = err && typeof err.code === "number" ? err.code : -32603;
    const emsg = err ? String(err.message ?? "") : "";
    if (p.kind === "single") {
      if (p.timer) clearTimeout(p.timer);
      if (err) {
        p.reject(new RpcServerError(code, emsg));
      } else {
        const r = msg.result;
        p.resolve(typeof r === "object" && r !== null ? (r as Record<string, unknown>) : {});
      }
      return;
    }
    // stream — this is the final reply (finish_reason + usage, no content).
    p.chunks.close();
    if (err) {
      p.rejectFinal(new RpcServerError(code, emsg));
    } else {
      p.resolveFinal(parseLlmCompleteResult(msg.result));
    }
  }

  private routeDelta(params: unknown): void {
    const p = (typeof params === "object" && params !== null ? params : {}) as {
      request_id?: unknown;
      chunk?: unknown;
    };
    const pending = typeof p.request_id === "number" ? this.pending.get(p.request_id) : undefined;
    if (pending && pending.kind === "stream" && typeof p.chunk === "string") {
      pending.chunks.push(p.chunk);
    } else {
      process.stderr.write(
        `plugin: llm.complete.delta for unknown/non-stream request ${JSON.stringify(p.request_id)}, dropped\n`,
      );
    }
  }

  private abandonPending(): void {
    for (const [id, p] of [...this.pending.entries()]) {
      this.pending.delete(id);
      if (p.kind === "single") {
        if (p.timer) clearTimeout(p.timer);
        p.reject(new RpcTransportError("adapter shutting down"));
      } else {
        p.chunks.close();
        p.rejectFinal(new RpcTransportError("adapter shutting down"));
      }
    }
  }

  // ── lifecycle ───────────────────────────────────────────────────

  private replyInitialize(id: unknown): void {
    if (id === undefined) {
      return; // spec-violating notification; nothing to reply to.
    }
    const result: Record<string, unknown> = {
      manifest: this.parsed.raw,
      server_version: this.serverVersion,
    };
    if (this.declaredTools.length > 0) {
      result.tools = this.declaredTools.map(toolDefToJson);
    }
    this.writeFrame(buildResponse(id as never, result));
  }

  // ── tool.invoke dispatch (contract §5.t) ────────────────────────

  private handleToolInvoke(id: unknown, params: unknown): void {
    if (id === undefined || id === null) {
      process.stderr.write("plugin: tool.invoke frame without an id, dropped\n");
      return;
    }
    // No-handler wins over param shape — mirrors the Rust SDK so the
    // host's RemoteToolHandler surfaces a clear "not implemented".
    if (this.onTool === undefined && this.onToolWithContext === undefined) {
      this.writeFrame(buildErrorResponse(id as never, -32601, "method not found: tool.invoke"));
      return;
    }
    if (typeof params !== "object" || params === null) {
      this.writeFrame(buildErrorResponse(id as never, -32602, "tool.invoke params must be an object"));
      return;
    }
    const p = params as { plugin_id?: unknown; tool_name?: unknown; args?: unknown; agent_id?: unknown };
    if (typeof p.tool_name !== "string" || p.tool_name.length === 0) {
      this.writeFrame(buildErrorResponse(id as never, -32602, "tool.invoke params missing string `tool_name`"));
      return;
    }
    const inv: ToolInvocation = {
      pluginId: typeof p.plugin_id === "string" ? p.plugin_id : this.parsed.plugin.id,
      toolName: p.tool_name,
      args: p.args ?? null,
      agentId: typeof p.agent_id === "string" ? p.agent_id : undefined,
    };
    const task: Promise<void> = Promise.resolve()
      .then(() => this.invokeToolHandler(inv))
      .then((result) => {
        this.sendToolResult(id, result);
      })
      .catch((e) => {
        if (e instanceof ToolError) {
          this.sendToolError(id, e);
        } else {
          this.sendToolError(
            id,
            new ToolExecutionFailedError(e instanceof Error ? e.message : String(e)),
          );
        }
      });
    this.inflight.add(task);
    task.finally(() => {
      this.inflight.delete(task);
    });
  }

  private invokeToolHandler(inv: ToolInvocation): unknown | Promise<unknown> {
    if (this.onToolWithContext !== undefined) {
      const ctx: ToolContext = { broker: this.broker, pluginId: this.parsed.plugin.id };
      return this.onToolWithContext(inv, ctx);
    }
    // onTool is defined — checked in handleToolInvoke before this runs.
    return (this.onTool as ToolHandler)(inv);
  }

  private sendToolResult(id: unknown, result: unknown): void {
    let line: string;
    try {
      line = serializeFrame(buildResponse(id as never, result));
    } catch (e) {
      this.sendToolError(
        id,
        new ToolExecutionFailedError(
          `tool result not JSON-serializable: ${e instanceof Error ? e.message : String(e)}`,
        ),
      );
      return;
    }
    process.stdout.write(line);
  }

  private sendToolError(id: unknown, err: ToolError): void {
    const data = err.errorData();
    const errorObj: { code: number; message: string; data?: unknown } = {
      code: err.code,
      message: err.message,
    };
    if (data !== undefined) errorObj.data = data;
    const frame: JsonRpcErrorResponse = {
      jsonrpc: JSONRPC_VERSION,
      id: id as JsonRpcId,
      error: errorObj,
    };
    this.writeFrame(frame);
  }

  private async replyShutdown(id: unknown): Promise<void> {
    this.abandonPending();
    await this.drainInflight();
    if (this.onShutdown !== undefined) {
      try {
        await this.onShutdown();
      } catch (e) {
        process.stderr.write(`plugin: onShutdown raised: ${e instanceof Error ? e.message : String(e)}\n`);
      }
    }
    if (id !== undefined) {
      this.writeFrame(buildResponse(id as never, { ok: true }));
    }
  }

  private dispatchEvent(params: unknown): void {
    if (this.onEvent === undefined) {
      return;
    }
    let topic: string;
    let event: Event;
    try {
      if (typeof params !== "object" || params === null) {
        throw new WireError("broker.event params must be a JSON object");
      }
      const p = params as { topic?: unknown; event?: unknown };
      if (typeof p.topic !== "string") {
        throw new WireError("broker.event params missing string `topic`");
      }
      topic = p.topic;
      event = Event.fromJson(p.event);
    } catch (e) {
      process.stderr.write(`plugin: dispatch decode failed: ${e instanceof Error ? e.message : String(e)}\n`);
      return;
    }

    const handler = this.onEvent;
    const task: Promise<void> = Promise.resolve()
      .then(() => handler(topic, event, this.broker))
      .catch((e) => {
        process.stderr.write(`plugin: onEvent raised: ${e instanceof Error ? e.message : String(e)}\n`);
      });
    this.inflight.add(task);
    task.finally(() => {
      this.inflight.delete(task);
    });
  }

  private writeFrame(frame: Parameters<typeof serializeFrame>[0]): void {
    process.stdout.write(serializeFrame(frame));
  }

  private async drainInflight(): Promise<void> {
    if (this.inflight.size === 0) {
      return;
    }
    await Promise.allSettled([...this.inflight]);
  }
}

// Re-export ManifestError so importers can tell it apart from generic
// PluginError without pulling errors.js too.
export { ManifestError } from "./errors.js";

// Touch helpers/types referenced only for side effects / type-position
// so the linter doesn't flag the imports as unused.
void isStdoutGuardInstalled;
void JSONRPC_VERSION;
